src/node/blockstorage.cpp

```diff
@@ -29,6 +30,7 @@ std::atomic_bool fImporting(false);
 std::atomic_bool fReindex(false);
 bool fPruneMode = false;
 uint64_t nPruneTarget = 0;
+std::shared_mutex cs_UnlinkFiles;
```

Comment:
```suggestion
std::shared_mutex g_unlinkfiles_mutex;
```

Would be more inline with our naming conventions.

But generally I think we should be removing globals instead of adding more. In this case, we could make `ReadBlockFromDisk`, `ReadRawBlockFromDisk`, `UndoReadFromDisk`, `UnlinkPrunedFiles` methods of `BlockManager` and have the mutex owned by that as well. I have not scoped out how big of a refactor that would be but I think that would be preferable.

Reply:
All places have access to the `BlockManager` fairly easily so it should be straightforward refactor, with the exception of `CZMQPublishRawBlockNotifier::NotifyBlock`. I have no idea how I will get a handle to the `BlockManager` from there :/.

Reply:
I created https://github.com/bitcoin/bitcoin/pull/26375 to remove needing to read a block from `CZMQPublishRawBlockNotifier::NotifyBlock` entirely. After that or something like it the refactor can be done much more easily. Marking as draft until that is merged.

Reply:
The refactor should be after and separate from the fix.